massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+//! The speculative ledger represents, in a compressed way,
+//! the state of the ledger at an arbitrary execution slot.
+//! It never actually writes to the consensus state
+//! but keeps track of the changes that were applied to it since its creation.
+
+use massa_execution_exports::ExecutionError;
+use massa_hash::hash::Hash;
+use massa_ledger::{Applicable, FinalLedger, LedgerChanges};
+use massa_models::{Address, Amount};
+use parking_lot::RwLock;
+use std::sync::Arc;
+
+/// The SpeculativeLedger contains an thread-safe shared reference to the final ledger (read-only),
+/// a list of existing changes that happened o the ledger since its finality,
+/// as well as an extra list of "added" changes.
+/// The SpeculativeLedger makes it possible to transparently manipulate a virtual ledger
+/// that takes into account all those ledger changes and allows adding more
+/// while keeping track of all the newly added changes, and never writing in the final ledger.
+pub struct SpeculativeLedger {
+    /// Thread-safe shared access to the final ledger. For reading only.
+    final_ledger: Arc<RwLock<FinalLedger>>,
+
+    /// Accumulation of changes that previously happened to the ledger since finality.
+    /// This value is not modified by changes applied to the SpeculativeLedger.
+    ///
+    /// TODO maybe have the history directly here,
```

Comment:
is there an issue for that ?

Reply:
there will be one if we decide that this approach is what we want

Reply:
Don't we need an issue to have a discussion to make that decision?

Reply:
I made a followup issue to discuss this potential improvement https://github.com/massalabs/massa/issues/2343